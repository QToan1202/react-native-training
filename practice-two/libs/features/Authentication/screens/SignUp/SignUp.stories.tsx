import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import SignUp from './index'

export default {
  title: 'screens/SignUp',
  component: SignUp,
} as ComponentMeta<typeof SignUp>

const Template: ComponentStory<typeof SignUp> = () => <SignUp />

export const Default = Template.bind({})
Default.args = {}
