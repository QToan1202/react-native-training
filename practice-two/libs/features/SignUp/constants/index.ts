// eslint-disable-next-line import/prefer-default-export
export { default as SIGN_UP_INPUTS } from './signUpFields'
