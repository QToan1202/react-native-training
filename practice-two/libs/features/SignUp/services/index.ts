import { IUser, add, get } from '@practice-two/shared'

export const register = async (path: string, user: IUser): Promise<IUser> => {
  const { email } = user
  const users: IUser[] = await get(path, { params: { email } })

  // Check register Email exist
  if (users.length) throw Error('This email already use by other client') // TODO: Extract to error message constant

  return add<IUser>(path, user)
}

export default register
