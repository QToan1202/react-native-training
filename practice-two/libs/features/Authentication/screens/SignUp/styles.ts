import { StyleSheet } from 'react-native'

import { COLORS } from '@practice-two/shared/constants'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.PRIMARY,
    paddingHorizontal: 32,
    justifyContent: 'center',
  },
})

export default styles
