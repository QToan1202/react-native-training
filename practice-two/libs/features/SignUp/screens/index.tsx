/* eslint-disable react-hooks/exhaustive-deps */
import { useCallback, useMemo } from 'react'
import { KeyboardAvoidingView, Platform } from 'react-native'
import { SubmitHandler, useForm } from 'react-hook-form'
import { SafeAreaView } from 'react-native-safe-area-context'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { XStack, YStack } from 'tamagui'

import { Button, Heading, Input, Paragraph, IForm, IUser, Feature } from '@practice-two/shared'
import { RootStackParamList } from 'libs/shared/navigation/Stack'
import { useRegister } from '../hooks'
import { SIGN_UP_INPUTS } from '../constants'

import styles from './styles'

export type SignUpScreenProps = NativeStackScreenProps<RootStackParamList, 'SignUp'>

const SignUp = ({ navigation }: SignUpScreenProps) => {
  const { mutate, status } = useRegister(process.env.USER_ENDPOINT)
  const { control, handleSubmit, watch } = useForm<IForm>()
  const observePassword: string = watch('password')
  const onSubmit: SubmitHandler<IForm> = useCallback((data: IUser) => {
    mutate(data)
  }, [])
  const handleToLoginScreen = useCallback(() => navigation.navigate('Login'), [])
  const SignUpInputs = useMemo(
    () => (
      <YStack space>
        {SIGN_UP_INPUTS(observePassword).map(({ name, ...props }) => (
          <Input key={name} {...props} control={control} name={name} />
        ))}
      </YStack>
    ),
    [control, observePassword]
  )

  return (
    <Feature feat="signup">
      <SafeAreaView style={styles.container}>
        <YStack marginBottom="$space.5" space="$space.9" alignItems="center">
          <Heading content="Welcome to tradly" fontWeight="$2" letterSpacing="$2" />
          <Paragraph content="Signup to your account" fontSize="$2" letterSpacing="$4" />
        </YStack>
        <KeyboardAvoidingView behavior={Platform.OS === 'ios' ? 'padding' : 'height'}>
          {SignUpInputs}
        </KeyboardAvoidingView>
        <Button
          title="create"
          variant="secondary"
          marginVertical="$space.7"
          letterSpacing="$4"
          fontSize="$2"
          isDisable={status === 'pending'}
          onPress={handleSubmit(onSubmit)}
        />
        <XStack justifyContent="center" alignItems="baseline" space="$space.1">
          <Paragraph content="Have an account?" fontSize="$3" lineHeight="$4" />
          <Button
            title="Sign in"
            variant="tertiary"
            borderColor="$color.transparent"
            fontWeight="$3"
            fontSize="$3"
            lineHeight="$4"
            onPress={handleToLoginScreen}
          />
        </XStack>
      </SafeAreaView>
    </Feature>
  )
}

export default SignUp
