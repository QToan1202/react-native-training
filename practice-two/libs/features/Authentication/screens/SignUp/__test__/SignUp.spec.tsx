import { render } from '@testing-library/react-native'

import SignUp from '../index'

describe('Testing Sign Up screen', () => {
  it('Render screen successfully', () => {
    const component = render(<SignUp />).toJSON()

    expect(component).toMatchSnapshot()
  })
})
