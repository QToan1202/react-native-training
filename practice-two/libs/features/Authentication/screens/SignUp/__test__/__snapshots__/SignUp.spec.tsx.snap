// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Sign Up screen Render screen successfully 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#33907C",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 32,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "marginBottom": 25,
        "rowGap": 54,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_700Bold",
            "fontSize": 24,
          },
          {
            "fontFamily": "Montserrat_500Medium",
          },
          undefined,
        ]
      }
    >
      Welcome to tradly
    </Text>
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_400Regular",
          },
          {
            "fontSize": 16,
          },
          undefined,
        ]
      }
    >
      Signup to your account
    </Text>
  </View>
  <View
    onLayout={[Function]}
    style={
      {
        "paddingBottom": 0,
      }
    }
  >
    <View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="First Name"
        placeholderTextColor="#fff"
        style={
          [
            {
              "borderColor": "#fff",
              "borderRadius": 50,
              "borderWidth": 1,
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
              "fontSize": 18,
              "paddingLeft": 15,
              "paddingVertical": 12,
            },
            {},
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          {
            "height": 20,
          }
        }
      />
    </View>
    <View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Last Name"
        placeholderTextColor="#fff"
        style={
          [
            {
              "borderColor": "#fff",
              "borderRadius": 50,
              "borderWidth": 1,
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
              "fontSize": 18,
              "paddingLeft": 15,
              "paddingVertical": 12,
            },
            {},
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          {
            "height": 20,
          }
        }
      />
    </View>
    <View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Email/Mobile Number"
        placeholderTextColor="#fff"
        style={
          [
            {
              "borderColor": "#fff",
              "borderRadius": 50,
              "borderWidth": 1,
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
              "fontSize": 18,
              "paddingLeft": 15,
              "paddingVertical": 12,
            },
            {},
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          {
            "height": 20,
          }
        }
      />
    </View>
    <View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Password"
        placeholderTextColor="#fff"
        secureTextEntry={true}
        style={
          [
            {
              "borderColor": "#fff",
              "borderRadius": 50,
              "borderWidth": 1,
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
              "fontSize": 18,
              "paddingLeft": 15,
              "paddingVertical": 12,
            },
            {},
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          {
            "height": 20,
          }
        }
      />
    </View>
    <View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Re-enter password"
        placeholderTextColor="#fff"
        secureTextEntry={true}
        style={
          [
            {
              "borderColor": "#fff",
              "borderRadius": 50,
              "borderWidth": 1,
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
              "fontSize": 18,
              "paddingLeft": 15,
              "paddingVertical": 12,
            },
            {},
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          {
            "height": 20,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "transparent",
        "borderRadius": 50,
        "borderWidth": 2,
        "marginVertical": 40,
        "opacity": 1,
        "paddingVertical": 13,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "columnGap": 6,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "color": "#33907C",
              },
              {
                "fontFamily": "Montserrat_500Medium",
                "textAlign": "center",
                "textTransform": "capitalize",
              },
              undefined,
            ],
          ]
        }
      >
        create
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "baseline",
        "columnGap": 3,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_400Regular",
          },
          {
            "fontSize": 18,
          },
          undefined,
        ]
      }
    >
      Have an account?
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "transparent",
          "borderColor": "transparent",
          "borderRadius": 50,
          "borderWidth": 2,
          "opacity": 1,
          "paddingVertical": 13,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "columnGap": 6,
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "color": "#fff",
                },
                {
                  "fontFamily": "Montserrat_500Medium",
                  "textAlign": "center",
                  "textTransform": "capitalize",
                },
                {
                  "fontFamily": "Montserrat_600SemiBold",
                },
              ],
            ]
          }
        >
          Sign in
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
